sql: disallow dropping column referenced in partial index predicate

mgartner opened the pull request:
Informs #96924

Release note (bug fix/sql change): Columns referenced in partial index
predicates and partial unique constraint predicates can no longer be
dropped. The `ALTER TABLE .. DROP COLUMN` statement now returns an error
with a hint suggesting to drop the indexes and constraints first. This
is a temporary safe-guard to prevent users from hitting #96924. This
restriction will be lifted when that bug is fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97372)
<!-- Reviewable:end -->

mgartner commented:
This _should_ temporarily address #97370 by disallowing the column to be dropped.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-:-NOkywubDwtIhLNctVco:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1961 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOky8mb1S_dOhFgCY4I:-NOky8mb1S_dOhFgCY4J:b-6imqgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/logictest/testdata/logic_test/partial_index#L1961)):*
> ```
> 
> statement error column "g" cannot be dropped because it is referenced by partial index "t96924_g_j_idx"
> ALTER TABLE t96924 DROP COLUMN g
> ```

what about adding a test for column j?


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go:
```diff
@@ -353,7 +353,18 @@ func walkDropColumnDependencies(b BuildCtx, col *scpb.Column, fn func(e scpb.Ele
 				for _, columnID := range elt.EmbeddedExpr.ReferencedColumnIDs {
 					if columnID == col.ColumnID {
-						fn(e)
+						indexElts := b.QueryByID(elt.TableID).Filter(hasIndexIDAttrFilter(elt.IndexID))
```
maybe add a helper function like this to help (assuming that we only want to panic if the index predicate references the column, and it's ok that the index columns contain the column):
```
func maybePanicOnReferencingPartialIndex(b BuildCtx, col *scpb.Column, tblElts ElementResultSet) {
	var referencingIndexes catid.IndexSet
	tblElts.ForEachElementStatus(func(_ scpb.Status, _ scpb.TargetStatus, e scpb.Element) {
		switch elt := e.(type) {
		case *scpb.SecondaryIndex:
			if elt.EmbeddedExpr != nil {
				for _, columnID := range elt.EmbeddedExpr.ReferencedColumnIDs {
					if columnID == col.ColumnID {
						referencingIndexes.Add(elt.IndexID)
					}
				}
			}
		case *scpb.SecondaryIndexPartial:
			for _, columnID := range elt.ReferencedColumnIDs {
				if columnID == col.ColumnID {
					referencingIndexes.Add(elt.IndexID)
				}
			}
		}
	})

	if referencingIndexes.Len() > 0 {
		indexNames := make([]string, 0, referencingIndexes.Len())
		var colName string
		tblElts.ForEachElementStatus(func(_ scpb.Status, _ scpb.TargetStatus, e scpb.Element) {
			switch elt := e.(type) {
			case *scpb.IndexName:
				if referencingIndexes.Contains(elt.IndexID) {
					indexNames = append(indexNames, elt.Name)
				}
			case *scpb.ColumnName:
				if elt.ColumnID == col.ColumnID {
					colName = elt.Name
				}
			}
		})
		panic(errors.WithHint(
			pgerror.Newf(
				pgcode.InvalidColumnReference,
				"column %q cannot be dropped because it is referenced by partial index (%v)",
				colName, indexNames,
			),
			"drop the partial index first, then drop the column",
		))
	}
}
```

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-:-NOlNS5CAVUTXrNlHvJN:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/alter_table.go` line 1660 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOlKRaCA5yu9EX_zXCf:-NOlKRaCA5yu9EX_zXCg:b-218hxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/alter_table.go#L1660)):*
> ```Go
> 						colToDrop.ColName(), idx.GetName(),
> 					),
> 					"drop the partial index first, then drop the column",
> ```

[nit] Should we link to the issue?
___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1920 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOlL7sg1FjR0XgU03iG:-NOlL7sg1FjR0XgU03iH:bsrcsxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/logictest/testdata/logic_test/partial_index#L1920)):*
> ```
> subtest column_dropped_referenced_in_partial_index
> 
> # TODO(mgartner): Add UNIQUE WITHOUT INDEX.
> ```

[nit] Is this TODO outdated?
___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOlLJF17Q8yRlFZMh5Y:-NOlLJF17Q8yRlFZMh5Z:b-skqhoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/logictest/testdata/logic_test/partial_index#L1984)):*
> ```
> 
> statement error column "g" cannot be dropped because it is referenced by partial index "t96924_g_j_idx"
> ALTER TABLE t96924 DROP COLUMN g
> ```

It might be nice to demonstrate that dropping the partial index allows dropping a column. Another edge case to consider is altering the primary key when (a) the previous PK was rowid and (b) there is a partial index referencing rowid.


<!-- Sent from Reviewable.io -->

rytaft commented:
I realize this PR doesn't cover the issue with user defined types. Can we add that too?





<!-- Sent from Reviewable.io -->

DrewKimball commented:
Sounds like @Xiang-Gu will be taking over this one.

Xiang-Gu committed:
sql: disallow dropping column referenced in partial index predicate

Informs #96924

Release note (bug fix/sql change): Columns referenced in partial index
predicates and partial unique constraint predicates can no longer be
dropped. The `ALTER TABLE .. DROP COLUMN` statement now returns an error
with a hint suggesting to drop the indexes and constraints first. This
is a temporary safe-guard to prevent users from hitting #96924. This
restriction will be lifted when that bug is fixed.

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go:
```diff
@@ -349,15 +351,16 @@ func walkDropColumnDependencies(b BuildCtx, col *scpb.Column, fn func(e scpb.Ele
 			}
 		case *scpb.SecondaryIndex:
```
I think we still want to check `SecondaryIndexPartial` given that in mixed version and before we fully migrated to `V23_1_SchemaChangerDeprecatedIndexPredicates` we still emit `SecondaryIndexPartial`, see https://github.com/cockroachdb/cockroach/blob/dda41f8bf34e20813a066c59840044530945ded4/pkg/sql/schemachanger/scdecomp/decomp.go#L560-L560

DrewKimball reviewed (commented):
Looks pretty good to me, no comments beyond the ones I made earlier.

> Reviewed 3 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-:-NP-ysl0DkNy6n4GuGfP:ba9dtn)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for taking this over! Would it be very difficult to also address https://github.com/cockroachdb/cockroach/issues/96337 in this PR (prevent dropping user-defined types that are referenced by a partial index)?





<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97372)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @chengxiong-ruan, @DrewKimball, and @rytaft)

___
*[`pkg/sql/alter_table.go` line 1660 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOlKRaCA5yu9EX_zXCf:-NP3LEYHE1FeADiic9Dk:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/alter_table.go#L1660)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Should we link to the issue?
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOkuGcysJCrSysM4wMj-r1-356:-NP3X8Ro2427V0AeCUCK:bnc21lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L356)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

maybe add a helper like this to help and panic at the very beginning of this function instead of here. (assuming that we only want to panic if the index predicate references the column, and it's ok that the index columns contain the column):
Note that need to check `SecondaryIndexPartial` since we just migrated it into the `SecondaryIndex` for 23.1
```
func maybePanicOnReferencingPartialIndex(b BuildCtx, col *scpb.Column, tblElts ElementResultSet) {
	var referencingIndexes catid.IndexSet
	tblElts.ForEachElementStatus(func(_ scpb.Status, _ scpb.TargetStatus, e scpb.Element) {
		switch elt := e.(type) {
		case *scpb.SecondaryIndex:
			if elt.EmbeddedExpr != nil {
				for _, columnID := range elt.EmbeddedExpr.ReferencedColumnIDs {
					if columnID == col.ColumnID {
						referencingIndexes.Add(elt.IndexID)
            break
					}
				}
			}
		case *scpb.SecondaryIndexPartial:
			for _, columnID := range elt.ReferencedColumnIDs {
				if columnID == col.ColumnID {
					referencingIndexes.Add(elt.IndexID)
          break
				}
			}
		}
	})

	if referencingIndexes.Len() > 0 {
		indexNames := make([]string, 0, referencingIndexes.Len())
		var colName string
		tblElts.ForEachElementStatus(func(_ scpb.Status, _ scpb.TargetStatus, e scpb.Element) {
			switch elt := e.(type) {
			case *scpb.IndexName:
				if referencingIndexes.Contains(elt.IndexID) {
					indexNames = append(indexNames, elt.Name)
				}
			case *scpb.ColumnName:
				if elt.ColumnID == col.ColumnID {
					colName = elt.Name
				}
			}
		})
		panic(errors.WithHint(
			pgerror.Newf(
				pgcode.InvalidColumnReference,
				"column %q cannot be dropped because it is referenced by partial index (%v)",
				colName, indexNames,
			),
			"drop the partial index first, then drop the column",
		))
	}
}
```

</blockquote></details>

Thanks for the hint. I've done similar to what you suggested here.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 352 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOkuGcysJCrSysM4wMj-r2-352:-NP3M15H4lgaa4vAwitC:b-dw7ral) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66cd01741ecaebe38f82e9b48fdab8db7c1b300/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L352)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

I think we still want to check `SecondaryIndexPartial` given that in mixed version and before we fully migrated to `V23_1_SchemaChangerDeprecatedIndexPredicates` we still emit `SecondaryIndexPartial`, see https://github.com/cockroachdb/cockroach/blob/dda41f8bf34e20813a066c59840044530945ded4/pkg/sql/schemachanger/scdecomp/decomp.go#L560-L560
</blockquote></details>

Good point! Done



_Code quote:_
```Go
Temporarily disallow dropping a colum
```
___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1920 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOlL7sg1FjR0XgU03iG:-NP3XDv39owS3-UeZp12:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/logictest/testdata/logic_test/partial_index#L1920)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Is this TODO outdated?
</blockquote></details>

ack
___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1961 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOky8mb1S_dOhFgCY4I:-NP3XF8W7_XxT5S6TqqX:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/logictest/testdata/logic_test/partial_index#L1961)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what about adding a test for column j?
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-NOlLJF17Q8yRlFZMh5Y:-NP3kFoH310M95N2OOHu:b-wvlc2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/07d45e959470cbcf645d0e5e2cac132704b91a23/pkg/sql/logictest/testdata/logic_test/partial_index#L1984)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It might be nice to demonstrate that dropping the partial index allows dropping a column. Another edge case to consider is altering the primary key when (a) the previous PK was rowid and (b) there is a partial index referencing rowid.
</blockquote></details>

done. But note that this behavior (of dropping `rowid` when `alter primary key`) is only in the declarative schema changer but not in the legacy schema changer. It is too hard to implement it in the legacy schema changer.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work! I'll defer final approval to @chengxiong-ruan.

> Reviewed 18 of 18 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97372#-:-NP3yDdZA0_LxCOZE_6_:bt1q0k1)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan and @rytaft)



<!-- Sent from Reviewable.io -->

Xiang-Gu commented:
TFTR!

bors r+

Xiang-Gu commented:
bors r-

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go:
```diff
@@ -391,6 +390,56 @@ func walkDropColumnDependencies(b BuildCtx, col *scpb.Column, fn func(e scpb.Ele
+		switch elt := e.(type) {
+		case *scpb.SecondaryIndex:
+			if elt.EmbeddedExpr != nil && contains(elt.EmbeddedExpr.ReferencedColumnIDs, col.ColumnID) {
+				violatingIndex = elt.IndexID
```
nit: looks like we may have multiple violating indexes, but we only panic only the last one?

Xiang-Gu replied:
This is only for the error message, and I think it's okay that we only include one.

chengxiong-ruan replied:
sgtm

chengxiong-ruan reviewed (approved):
LGTM mod one nit. Nice work!

Xiang-Gu commented:
bors r+

Xiang-Gu commented:
bors r+

Xiang-Gu commented:
bors r- 

I found an edge case that is not well covered. Let me amend it.

Xiang-Gu commented:
Added some test cases where to-be-dropped column appears in the predicate but not in the keyed columns (e.g. INDEX (i) WHERE (j > 0), DROP COLUMN j). Ready to bors again.

bors r+

rytaft commented:
FYI to all -- in the future, please use a single category for release notes. The release notes script does not understand "Release note (bug fix/sql change)":
https://cockroachlabs.slack.com/archives/C0306LF0W69/p1677510172252459

mgartner commented:
Thanks for all the assistance on this one, everyone!
